staging/src/k8s.io/apimachinery/pkg/util/diff/diff.go

```diff
@@ -50,219 +47,83 @@ func StringDiff(a, b string) string {
-			fmt.Sprintf("  a: %s", elidedA),
-			fmt.Sprintf("  b: %s", elidedB),
-		)
+// ExpectEqual is a test helper that checks that objects, a and b, are equal.
+// Otherwise, it reports an error to the test interface with the diff.
+func ExpectEqual(t Testing, a, b interface{}, opts ...cmp.Option) {
+	t.Helper()
+	if !cmp.Equal(a, b, opts...) {
+		t.Errorf("unexpected difference:\n%s", cmp.Diff(a, b, opts...))
 	}
-	return strings.Join(out, "\n")
 }
 
-// limit:
-// 1. stringifies aObj and bObj
-// 2. elides identical prefixes if either is too long
-// 3. elides remaining content from the end if either is too long
-func limit(aObj, bObj interface{}, max int) (string, string) {
-	elidedPrefix := ""
-	elidedASuffix := ""
-	elidedBSuffix := ""
-	a, b := fmt.Sprintf("%#v", aObj), fmt.Sprintf("%#v", bObj)
-
-	if aObj != nil && bObj != nil {
-		if aType, bType := fmt.Sprintf("%T", aObj), fmt.Sprintf("%T", bObj); aType != bType {
-			a = fmt.Sprintf("%s (%s)", a, aType)
-			b = fmt.Sprintf("%s (%s)", b, bType)
-		}
-	}
-
-	for {
-		switch {
-		case len(a) > max && len(a) > 4 && len(b) > 4 && a[:4] == b[:4]:
-			// a is too long, b has data, and the first several characters are the same
-			elidedPrefix = "..."
-			a = a[2:]
-			b = b[2:]
-
-		case len(b) > max && len(b) > 4 && len(a) > 4 && a[:4] == b[:4]:
-			// b is too long, a has data, and the first several characters are the same
-			elidedPrefix = "..."
-			a = a[2:]
-			b = b[2:]
-
-		case len(a) > max:
-			a = a[:max]
-			elidedASuffix = "..."
-
-		case len(b) > max:
-			b = b[:max]
-			elidedBSuffix = "..."
-
-		default:
-			// both are short enough
-			return elidedPrefix + a + elidedASuffix, elidedPrefix + b + elidedBSuffix
-		}
-	}
+func legacyDiff(a, b interface{}) string {
+	return cmp.Diff(a, b, DeepAllowUnexported(a, b))
 }
 
-func public(s string) bool {
-	if len(s) == 0 {
-		return false
-	}
-	return s[:1] == strings.ToUpper(s[:1])
+// ObjectDiff prints the diff of two go objects.
+// DEPRECATED: use github.com/google/go-cmp/cmp.Diff
+func ObjectDiff(a, b interface{}) string {
+	return legacyDiff(a, b)
 }
 
-type diff struct {
-	path *field.Path
-	a, b interface{}
+// ObjectGoPrintDiff prints the diff of two go objects.
+// DEPRECATED: use github.com/google/go-cmp/cmp.Diff
+func ObjectGoPrintDiff(a, b interface{}) string {
+	return legacyDiff(a, b)
 }
 
-type orderedDiffs []diff
+// ObjectReflectDiff prints the diff of two go objects.
+// DEPRECATED: use github.com/google/go-cmp/cmp.Diff
+func ObjectReflectDiff(a, b interface{}) string {
+	return legacyDiff(a, b)
+}
 
-func (d orderedDiffs) Len() int      { return len(d) }
-func (d orderedDiffs) Swap(i, j int) { d[i], d[j] = d[j], d[i] }
-func (d orderedDiffs) Less(i, j int) bool {
-	a, b := d[i].path.String(), d[j].path.String()
-	if a < b {
-		return true
+// DeepAllowUnexporeted creates a cmp.Option that recursively allows unexported
+// fields in the passed in types.
+func DeepAllowUnexported(vs ...interface{}) cmp.Option {
```

Comment:
I think I'm going to try to add this to cmp/cmpoptions before merging this here.